### PR TITLE
Make the previous and next buttons always work

### PR DESCRIPTION
**Pull request by bergice**
Resolves https://github.com/silverstripe/silverstripe-admin/issues/886

### CODE REVIEW

**Comment by bergice** [comment]
Resolves https://github.com/silverstripe/silverstripe-admin/issues/886

**Review by emteknetnz — CHANGES_REQUESTED** [review]
Could you add some unit tests please

**Comment by bergice** [comment]
> Could you add some unit tests please

Spent some time trying to get a unit test done but it's quite tricky because of the convoluted state logic and pagination that's required.

A behat test would be more ideal but we don't have anything suited in `framework`.

**Comment by emteknetnz** [comment]
I don't feel confident merging this as is, or at least I don't feel confident that it won't break again sometime in the future.

We've assigned 3 story points on the parent issue so we've got some time to fix this.

Seems like we could add a corresponding behat test in admin to cover this functionality

File: src/Forms/GridField/GridFieldDetailForm_ItemRequest.php [hunk 1]
```diff
@@ -564,27 +565,20 @@ public function getEditLink($id)
...
-        $limitOffset = max(0, $itemsPerPage * ($currentPage-1) -1);
-
-        $map = $list->limit($limit, $limitOffset)->column('ID');
+        $table = DataObject::getSchema()->tableForField($this->record, 'ID');
+        if ($offset > 0) {
+            $list = $list->where(["\"$table\".\"ID\" >= ?" => $this->record->ID])
+                ->sort(['ID' => 'ASC']);
+        } else {
+            $list = $list->where(["\"$table\".\"ID\" <= ?"  => $this->record->ID])
+                ->sort(['ID' => 'DESC']);
```
Comment (kinglozzer) [inline src/Forms/GridField/GridFieldDetailForm_ItemRequest.php:577]: Doesn’t this change the sorting completely? Any sort order applied to the gridfield (or a `default_sort`) will be ignored, and the records will be sorted by ID. Is that intended?

Comment (bergice) [reply src/Forms/GridField/GridFieldDetailForm_ItemRequest.php:577]: > Doesn’t this change the sorting completely? Any sort order applied to the gridfield (or a `default_sort`) will be ignored, and the records will be sorted by ID. Is that intended?

Hmm, I'm pretty sure `$list->getManipulatedList()` already applies all the filters.

**Review by emteknetnz — CHANGES_REQUESTED** [review]
This approach seems like it won't fix the root cause of the issue

I've done some [investigation](https://github.com/silverstripe/silverstripe-admin/issues/886#issuecomment-876813345) into part of the issue

This pull-request just filtering and sorting `$list`, however $list is fundamentally incomplete, which is why the next button can only be clicked a couple of times.  I did not investigate the other issue identified of results being in the incorrect order.  There I'd say it's likely the `$list` is simply incorrect

A proper solution would be to get a correct `$list`

If you're able to update the PR to get a correct `$list` that would still work in a 100K database records scenario, then great, otherwise recommend this one be closed

**Comment by bergice** [comment]
Closing, while the filtering is still working as expected, sorting is broken and will need to be fixed. Can reopen later if we want.

**Pull request closed without merging**